tests/purs/failing/RowOneMissingLabel.purs

```diff
@@ -0,0 +1,6 @@
+-- @shouldFailWith PropertyIsMissing
+module A where
+
+-- Currently only reports on the missing `c` label
```

Comment:
Do we want to do something about this as well? This case looks like it'll be handled in `TypeChecker.Subsumption`.

Reply:
Ahh, yeah we should.

Reply:
There's a similarly-named [`unifyRows` in Kinds.hs](https://github.com/purescript/purescript/blob/master/src/Language/PureScript/TypeChecker/Kinds.hs#L447-L464). Should that also be updated to this approach? I'm not sure what error messages it would affect, but I imagine it would be desirable.

Reply:
> Maybe not this exact approach, but might be good to parU the labels at least. I don't think you get here without doing some pretty advanced stuff; something like

I wanted to compare the before/after of using the same approach. Here's what I get:

<details>
<summary>Ryan's above code</summary>

```purs
-- @shouldFailWith KindsDoNotUnify
module RowKindTest where

import Type.Proxy (Proxy(..))

foreign import data Foo :: Record (a :: Type, b :: Type)
foreign import data Bar :: Record (a :: Int, c :: Type, d :: Type) -> Type

something :: Proxy (Bar Foo)
something = Proxy
```

</details>

<details>
<summary><code>master</code> error:</summary>

```
Error found:
in module [33mRowKindTest[0m
at tests/purs/failing/RowKindTest.purs:9:25 - 9:28 (line 9, column 25 - line 9, column 28)

  Could not match kind
  [33m      [0m
  [33m  Type[0m
  [33m      [0m
  with kind
  [33m     [0m
  [33m  Int[0m
  [33m     [0m

while matching label [33ma[0m
while checking that type [33mFoo[0m
  has kind [33m{ a :: Int [0m
           [33m, c :: Type[0m
           [33m, d :: Type[0m
           [33m}          [0m
while inferring the kind of [33mBar Foo[0m
while inferring the kind of [33mProxy (Bar Foo)[0m
in value declaration [33msomething[0m

See https://github.com/purescript/documentation/blob/master/errors/KindsDoNotUnify.md for more information,
or to contribute content related to this error.
```

</details>

<details>
<summary>This PR's output</summary>

```
Error found:
in module [33mRowKindTest[0m
at tests/purs/failing/RowKindTest.purs:9:25 - 9:28 (line 9, column 25 - line 9, column 28)

  Could not match kind
  [33m             [0m
  [33m  ( a :: Type[0m
  [33m  , b :: Type[0m
  [33m  )          [0m
  [33m             [0m
  with kind
  [33m             [0m
  [33m  ( a :: Int [0m
  [33m  , c :: Type[0m
  [33m  , d :: Type[0m
  [33m  )          [0m
  [33m             [0m

while checking that type [33mFoo[0m
  has kind [33m{ a :: Int [0m
           [33m, c :: Type[0m
           [33m, d :: Type[0m
           [33m}          [0m
while inferring the kind of [33mBar Foo[0m
while inferring the kind of [33mProxy (Bar Foo)[0m
in value declaration [33msomething[0m

See https://github.com/purescript/documentation/blob/master/errors/KindsDoNotUnify.md for more information,
or to contribute content related to this error.
```

</details>

I think this PR's output is better than `master`, even if this is very advanced.

Reply:
I've pushed the kind unification-related changes. Let's continue that conversation in [this comment's thread](https://github.com/purescript/purescript/pull/4421#discussion_r1289219183). I'm still working on the `checkProperties` reimplementation. My initial work caused issues with 3701, so I'm trying a different approach. I likely won't get to this again until tomorrow.

Reply:
I took a stab at `checkProperties` myself, as well as the other comments I made in my last review and a grab-bag of record-related error message improvements. I have mostly stayed away from the kind-unification piece aside from a refactoring that I hope you'll agree is a good idea; I think you've done more thinking about that side of things than I have.